tutorials/application_design/py/subscriber.py

```diff
@@ -45,71 +38,110 @@ def __init__(
+
+        def display_handler():
+            while not display_handler_sentinel.is_set():
+                with display_handler_condition:
+                    display_handler_condition.wait(5)
+                    dashboard_condition.trigger_value = True
+
+        display_thread = threading.Thread(target=display_handler)
+
+        waitset = dds.WaitSet()
+        waitset.attach_condition(new_position_condition)
+        waitset.attach_condition(dashboard_condition)
+
+        display_thread.start()
+        try:
+            while True:
+                waitset.dispatch()
+                dashboard_condition.trigger_value = False
+        except KeyboardInterrupt:
+            with display_handler_condition:
+                display_handler_sentinel.set()
+                display_handler_condition.notify_all()
+
+        display_thread.join()
+
+    def _new_position_string(self, condition):
+        string = str()
+        for sample, info in self._transit_reader.select().condition(condition).read():
+            if not info.valid:
+                continue
+
+            string += f"[INFO] Vehicle {sample.vehicle_vin}"
+            if sample.current_route and len(sample.current_route):
+                string += f" is enroute to {sample.current_route[-1]} from {sample.current_position}"
+            else:
+                string += (
+                    f" has arrived at its destination in {sample.current_position}"
                 )
-                for coord in data.reached_destinations:
-                    print(f"    - {coord}")
-            print()
-            await asyncio.sleep(0.5)
-
-    async def _metrics_app(self):
-        async for sample, info in self._metrics_reader.take_async():
-            if info.instance_handle not in self._dashboard_data:
+            string += "\n"
+        return string
+
+    def _dashboard_string(self):
+        dashboard_data = self._dashboard_data()
+        online_vehicles = [
+            data for data in dashboard_data.values() if not data.is_historical
+        ]
+        offline_vehicles = [
+            data for data in dashboard_data.values() if data.is_historical
+        ]
+
+        online_str = "\n".join(
+            f"Vehicle {data.vin}:\n"
+            f"  Fuel updates: {len(data.fuel_history)}\n"
+            f"  Last known destination: {data.current_destination}\n"
+            f"  Last known fuel level: {data.fuel_history[-1]}\n"
+            for data in online_vehicles
+        )
+        offline_str = "\n".join(f"Vehicle {data.vin}" for data in offline_vehicles)
+
+        return "\n".join(
+            [
+                f"[[ DASHBOARD: {datetime.now()} ]]",
+                f"Online vehicles: {len(online_vehicles)}",
+                online_str,
+                f"Offline vehicles: {len(offline_vehicles)}",
+                offline_str,
+            ]
+        )
+
+    def _dashboard_data(self):
+        data: typing.Dict[dds.InstanceHandle, DashboardItem] = {}
```

Comment:
If you use typing for a variable here, you should also use it for the parameters and return types.

Also, prefer `data: Dict[..., ...]`, and `from typing import Dict, ...`.

Reply:
Some type hints can be easily inferred, but creating an empty dictionary is not one of these scenarios. I'm limiting hinting to these scenarios instead of being explicit in every scenario.